src/cloudflare/internal/d1-api.ts

```diff
@@ -98,34 +156,50 @@ class D1Database {
     return exec.map(toArrayOfObjects);
   }
 
-  public async exec(query: string): Promise<D1ExecResult> {
-    const lines = query.trim().split('\n');
-    const _exec = await this._send('/execute', lines, [], 'NONE');
-    const exec = Array.isArray(_exec) ? _exec : [_exec];
-    const error = exec
-      .map((r) => {
-        return r.error ? 1 : 0;
-      })
-      .indexOf(1);
-    if (error !== -1) {
-      throw new Error(
-        `D1_EXEC_ERROR: Error in line ${error + 1}: ${lines[error]}: ${
-          exec[error]?.error
-        }`,
-        {
-          cause: new Error(
-            `Error in line ${error + 1}: ${lines[error]}: ${exec[error]?.error}`
-          ),
-        }
-      );
+  // Returns the latest commit token we received from all responses processed so far.
+  // It does not return constraints that might have be passed during the session creation.
+  public getCommitToken(): D1SessionCommitToken | null {
+    switch (this.commitTokenOrConstraint) {
+      // First to any replica, and then anywhere that satisfies the commit token.
+      case 'first-unconstrained':
+      // First to primary, and then anywhere that satisfies the commit token.
+      case 'first-primary':
+        return null;
+      default:
+        return this.commitTokenOrConstraint;
+    }
+  }
+
+  // fetch will append the commit token header to all outgoing fetch calls.
+  // The response headers are parsed automatically, extracting the commit token
+  // from the response headers and updating it through `_updateCommitToken(token)`.
+  public async fetch(
+    input: RequestInfo | URL,
+    init?: RequestInit
+  ): Promise<Response> {
+    // We append the commit token to all fetch queries.
+    const h = new Headers(init?.headers);
+
+    // We send either a constraint, or a commit token, and the eyeball worker will figure out
+    // what to do based on the value. This simulates the same flow as the REST API would behave too.
+    if (this.commitTokenOrConstraint) {
+      h.set(D1_SESSION_COMMIT_TOKEN_HTTP_HEADER, this.commitTokenOrConstraint);
+    }
+
+    if (!init) {
+      init = { headers: h };
     } else {
-      return {
-        count: exec.length,
-        duration: exec.reduce((p, c) => {
-          return p + (c.meta['duration'] as number);
-        }, 0),
-      };
+      init.headers = h;
     }
+    return this.fetcher.fetch(input, init).then((resp) => {
+      const newCommitToken = resp.headers.get(
+        D1_SESSION_COMMIT_TOKEN_HTTP_HEADER
+      );
+      if (newCommitToken) {
```

Comment:
It feels like we should require that if we send a commit token, we should expect a commit token in return and must raise an error if that's not the case.  Maybe drop a TODO?

Reply:
Hmm, good point. I haven't about it being required, but it makes sense to expect a token back if you are using a session, since otherwise the session semantics might break. I guess in the current state, the same token will be sent again next time, which will potentially "go back in time" in case there was any write afterwards.

It's a bit tricky during the rollout though of this change, since once this is merged there is always going to be a token sent (i.e. `first-primary`), but the backend API won't send anything back yet... 

I will put a TODO for now, and once the feature is fully rolled out we can add this check.

Reply:
I am inclined to add the validation later, since I don't know yet exactly what a token could look like or how we would validate it other than "is it not an empty string".

Also, this PR already adds a bit of logic changes, so don't want to introduce also a breaking change that requires eyeball changes to be merged and released first.